src/nvim/os/fileio.c

```diff
@@ -0,0 +1,319 @@
+    }
+  }
+}
+
+/// Read from file
+///
+/// @param[in,out]  fp  File to work with.
+/// @param[out]  ret_buf  Buffer to read to. Must not be NULL.
+/// @param[in]  size  Number of bytes to read. Buffer must have at least ret_buf
+///                   bytes.
+///
+/// @return error_code (< 0) or number of bytes read.
+ptrdiff_t file_read(FileDescriptor *const fp, char *const ret_buf,
+                    const size_t size)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
+{
+  assert(!fp->wr);
+  char *buf = ret_buf;
+  size_t read_remaining = size;
+  RBuffer *const rv = fp->rv;
+  while (read_remaining) {
+    const size_t rv_size = rbuffer_size(rv);
+    if (rv_size > 0) {
+      const size_t rsize = rbuffer_read(rv, buf, MIN(rv_size, read_remaining));
+      buf += rsize;
+      read_remaining -= rsize;
+    }
+    if (fp->eof) {
+      break;
+    }
+    if (read_remaining) {
+      assert(rbuffer_size(rv) == 0);
+      rbuffer_reset(rv);
+#ifdef HAVE_READV
+      // If there is readv() syscall, then take an opportunity to populate
+      // both target buffer and RBuffer at once, …
+      size_t write_count;
+      struct iovec iov[] = {
+        { .iov_base = buf, .iov_len = read_remaining },
+        { .iov_base = rbuffer_write_ptr(rv, &write_count),
+          .iov_len = kRWBufferSize },
+      };
+      assert(write_count == kRWBufferSize);
+      const ptrdiff_t r_ret = os_readv(fp->fd, &fp->eof, iov,
+                                       ARRAY_SIZE(iov));
+      if (r_ret > 0) {
+        if (r_ret > (ptrdiff_t)read_remaining) {
+          rbuffer_produced(rv, (size_t)(r_ret - (ptrdiff_t)read_remaining));
+          read_remaining = 0;
+        } else {
+          buf += (size_t)r_ret;
+          read_remaining -= (size_t)r_ret;
+        }
+      } else if (r_ret < 0) {
+        return r_ret;
+      }
+#else
+      if (read_remaining >= kRWBufferSize) {
+        // …otherwise leave RBuffer empty and populate only target buffer,
+        // because filtering information through rbuffer will be more syscalls.
+        const ptrdiff_t r_ret = os_read(fp->fd, &fp->eof, buf, read_remaining);
+        if (r_ret >= 0) {
+          read_remaining -= (size_t)r_ret;
+          return (ptrdiff_t)(size - read_remaining);
+        } else if (r_ret < 0) {
+          return r_ret;
+        }
+      } else {
+        size_t write_count;
+        const ptrdiff_t r_ret = os_read(fp->fd, &fp->eof,
+                                        rbuffer_write_ptr(rv, &write_count),
+                                        kRWBufferSize);
+        assert(write_count == kRWBufferSize);
+        if (r_ret > 0) {
+          rbuffer_produced(rv, (size_t)r_ret);
+        } else if (r_ret < 0) {
+          return r_ret;
+        }
+      }
+#endif
+    }
+  }
+  return (ptrdiff_t)(size - read_remaining);
+}
+
+/// Write to a file
+///
+/// @param[in]  fd  File descriptor to write to.
+/// @param[in]  buf  Data to write. May be NULL if size is zero.
```

Comment:
Source buffer for `rbuffer_write()` is expected to be not `NULL`. A size of zero also cannot be handled because `rbuffer_produced()` is called with that value.